packages/connect/src/base/listener.ts

```diff
@@ -377,24 +184,162 @@ class Listener extends EventEmitter implements InterfaceListener {
+    const waitForIncomingTcpPacket = defer<void>()
+
+    const TIMEOUT = 500
+
+    const abort = new AbortController()
+    const tcpTimeout = setTimeout(() => {
+      abort.abort()
+      waitForIncomingTcpPacket.reject()
+    }, TIMEOUT)
+    const udpTimeout = setTimeout(waitForIncomingUdpPacket.reject.bind(waitForIncomingUdpPacket), TIMEOUT)
+
+    const checkTcpMessage = (socket: TCPSocket) => {
+      socket.on('data', (data: Buffer) => {
+        if (u8aEquals(data, TCP_TEST)) {
+          clearTimeout(tcpTimeout)
+          waitForIncomingTcpPacket.resolve()
+        }
+      })
+    }
+    this.tcpSocket.on('connection', checkTcpMessage)
+
+    const checkUdpMessage = (msg: Buffer) => {
+      if (u8aEquals(msg, UDP_TEST)) {
+        clearTimeout(udpTimeout)
+        waitForIncomingUdpPacket.resolve()
+      }
+    }
+    this.udpSocket.on('message', checkUdpMessage)
 
-    const address = this.tcpSocket.address() as AddressInfo
+    const secondUdpSocket = createSocket('udp4')
+    secondUdpSocket.send(UDP_TEST, port, externalIp)
 
-    this.addrs.interface.push(
-      ...getAddrs(address.port, this.peerId.toB58String(), {
-        useIPv4: true,
-        includePrivateIPv4: true,
-        includeLocalhostIPv4: true
+    let done = false
+    const cleanUp = (): void => {
+      if (done) {
+        return
+      }
+      done = true
+      clearTimeout(tcpTimeout)
+      clearTimeout(udpTimeout)
+      this.udpSocket.removeListener('message', checkUdpMessage)
+      this.tcpSocket.removeListener('connection', checkTcpMessage)
+      tcpSocket.destroy()
+      secondUdpSocket.close()
+    }
+
+    const tcpSocket = createConnection({
+      port,
+      host: externalIp,
+      localPort: 9092,
+      signal: abort.signal
+    })
+      .on('connect', () => {
+        tcpSocket.write(TCP_TEST, (err: any) => {
+          if (err) {
+            log(`Failed to send TCP packet`, err)
+          }
+        })
       })
+      .on('error', (err: any) => {
+        if (err && (err.code == undefined || err.code !== 'ABORT_ERR')) {
+          error(`Error while checking NAT situation`, err.message)
+        }
+      })
+
+    if (!done) {
+      const results = await Promise.allSettled([waitForIncomingUdpPacket.promise, waitForIncomingTcpPacket.promise])
+
+      cleanUp()
+
+      return {
+        udpMapped: results[0].status === 'fulfilled',
+        tcpMapped: results[1].status === 'fulfilled'
+      }
+    }
+
+    return {
+      udpMapped: false,
+      tcpMapped: false
+    }
+  }
+
+  /**
+   * Attaches the listener to TCP and UDP sockets
+   * @param ma address to listen to
+   */
+  async listen(ma: Multiaddr): Promise<void> {
+    if (this.state == State.CLOSED) {
+      throw Error(`Cannot listen after 'close()' has been called`)
+    }
+
+    await this.bind(ma)
+
+    const ownInterface = this.tcpSocket.address() as AddressInfo
+
+    const natSituation = await this.checkNATSituation(ownInterface.address, ownInterface.port)
```

Comment:
bit more clever way to determine if we are running on an exposed host / behind a bidirectional NAT etc.